From 18750d206df928cf4f3f592da8c85569125b4386 Mon Sep 17 00:00:00 2001
Message-Id: <18750d206df928cf4f3f592da8c85569125b4386.1415496071.git.chang-joon.lee@intel.com>
In-Reply-To: <aa0c58fe046f9d00f3d381a75f6169b3f6d58c94.1415496071.git.chang-joon.lee@intel.com>
References: <aa0c58fe046f9d00f3d381a75f6169b3f6d58c94.1415496071.git.chang-joon.lee@intel.com>
From: Brad Volkin <bradley.d.volkin@intel.com>
Date: Wed, 21 May 2014 12:38:16 -0700
Subject: [PATCH 2/8] MUST_REBASE [VPG]: drm/i915: Additional restrictions on
 BBS commands

These checks limit 2nd-level batches to match what we expect from the
current Android userspace.

Marked as MUST_REBASE because upstream requirements here are different
due to behavior of already released userspace. Exact details are still
in discussion, but doing these checks is likely preferrable.

For: APDEV-870
Change-Id: I5bce0681f8ff1e79f50754274f62ae5e6c58bac4
Signed-off-by: Brad Volkin <bradley.d.volkin@intel.com>
---
 drivers/gpu/drm/i915/i915_cmd_parser.c |   26 +++++++++++++++++++++++++-
 drivers/gpu/drm/i915/i915_reg.h        |    2 ++
 2 files changed, 27 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/i915/i915_cmd_parser.c b/drivers/gpu/drm/i915/i915_cmd_parser.c
index 78085cb..1b3cab3 100644
--- a/drivers/gpu/drm/i915/i915_cmd_parser.c
+++ b/drivers/gpu/drm/i915/i915_cmd_parser.c
@@ -139,7 +139,6 @@ static const struct drm_i915_cmd_descriptor common_cmds[] = {
 			.mask = MI_GLOBAL_GTT,
 			.expected = 0,
 	      }},						       ),
-	CMD(  MI_BATCH_BUFFER_START,            SMI,   !F,  0xFF,   S  ),
 };
 
 static const struct drm_i915_cmd_descriptor render_cmds[] = {
@@ -202,6 +201,10 @@ static const struct drm_i915_cmd_descriptor render_cmds[] = {
 	      }},						       ),
 };
 
+static const struct drm_i915_cmd_descriptor ivb_render_cmds[] = {
+	CMD(  MI_BATCH_BUFFER_START,            SMI,   !F,  0xFF,   R  ),
+};
+
 static const struct drm_i915_cmd_descriptor hsw_render_cmds[] = {
 	CMD(  MI_SET_PREDICATE,                 SMI,    F,  1,      S  ),
 	CMD(  MI_RS_CONTROL,                    SMI,    F,  1,      S  ),
@@ -213,6 +216,18 @@ static const struct drm_i915_cmd_descriptor hsw_render_cmds[] = {
 	CMD(  MI_RS_STORE_DATA_IMM,             SMI,   !F,  0xFF,   S  ),
 	CMD(  MI_LOAD_URB_MEM,                  SMI,   !F,  0xFF,   S  ),
 	CMD(  MI_STORE_URB_MEM,                 SMI,   !F,  0xFF,   S  ),
+	CMD(  MI_BATCH_BUFFER_START,            SMI,   !F,  0xFF,   B,
+	      .bits = {{
+			.offset = 0,
+			.mask = (MI_BATCH_2ND_LEVEL_HSW |
+				 MI_BATCH_ADD_OFFSET_HSW |
+				 MI_BATCH_NON_SECURE_HSW |
+				 MI_BATCH_PPGTT_HSW),
+			.expected = (MI_BATCH_2ND_LEVEL_HSW |
+				     0 |
+				     MI_BATCH_NON_SECURE_HSW |
+				     MI_BATCH_PPGTT_HSW),
+	      }},                                                      ),
 	CMD(  GFX_OP_3DSTATE_DX9_CONSTANTF_VS,  S3D,   !F,  0x7FF,  S  ),
 	CMD(  GFX_OP_3DSTATE_DX9_CONSTANTF_PS,  S3D,   !F,  0x7FF,  S  ),
 
@@ -252,6 +267,12 @@ static const struct drm_i915_cmd_descriptor video_cmds[] = {
 			.condition_offset = 0,
 			.condition_mask = MI_FLUSH_DW_OP_MASK,
 	      }},						       ),
+	CMD(  MI_BATCH_BUFFER_START,            SMI,   !F,  0xFF,   B,
+	      .bits = {{
+			.offset = 0,
+			.mask = MI_BATCH_PPGTT_HSW,
+			.expected = MI_BATCH_PPGTT_HSW,
+	      }},                                                      ),
 	CMD(  MI_CONDITIONAL_BATCH_BUFFER_END,  SMI,   !F,  0xFF,   B,
 	      .bits = {{
 			.offset = 0,
@@ -295,6 +316,7 @@ static const struct drm_i915_cmd_descriptor vecs_cmds[] = {
 			.condition_offset = 0,
 			.condition_mask = MI_FLUSH_DW_OP_MASK,
 	      }},						       ),
+	CMD(  MI_BATCH_BUFFER_START,            SMI,   !F,  0xFF,   R  ),
 	CMD(  MI_CONDITIONAL_BATCH_BUFFER_END,  SMI,   !F,  0xFF,   B,
 	      .bits = {{
 			.offset = 0,
@@ -332,6 +354,7 @@ static const struct drm_i915_cmd_descriptor blt_cmds[] = {
 			.condition_offset = 0,
 			.condition_mask = MI_FLUSH_DW_OP_MASK,
 	      }},						       ),
+	CMD(  MI_BATCH_BUFFER_START,            SMI,   !F,  0xFF,   R  ),
 	CMD(  COLOR_BLT,                        S2D,   !F,  0x3F,   S  ),
 	CMD(  SRC_COPY_BLT,                     S2D,   !F,  0x3F,   S  ),
 };
@@ -356,6 +379,7 @@ static const struct drm_i915_cmd_descriptor hsw_blt_cmds[] = {
 static const struct drm_i915_cmd_table gen7_render_cmds[] = {
 	{ common_cmds, ARRAY_SIZE(common_cmds) },
 	{ render_cmds, ARRAY_SIZE(render_cmds) },
+	{ ivb_render_cmds, ARRAY_SIZE(ivb_render_cmds) },
 };
 
 static const struct drm_i915_cmd_table hsw_render_ring_cmds[] = {
diff --git a/drivers/gpu/drm/i915/i915_reg.h b/drivers/gpu/drm/i915/i915_reg.h
index 67475112..8763560 100644
--- a/drivers/gpu/drm/i915/i915_reg.h
+++ b/drivers/gpu/drm/i915/i915_reg.h
@@ -303,6 +303,8 @@
 #define   MI_BATCH_NON_SECURE_HSW	(1<<13)
 #define MI_BATCH_BUFFER_START	MI_INSTR(0x31, 0)
 #define   MI_BATCH_GTT		    (2<<6) /* aliased with (1<<7) on gen4 */
+#define   MI_BATCH_2ND_LEVEL_HSW    (1<<22)
+#define   MI_BATCH_ADD_OFFSET_HSW   (1<<16)
 #define MI_BATCH_BUFFER_START_GEN8	MI_INSTR(0x31, 1)
 
 
-- 
1.7.9.5

